Lambda-promtail: Add support for processing SQS messages, add promtailClient Type, add logger, upgrade dependencies and fix unexpected flushing behaviors (#8231)
grafana/loki: docs/sources/clients/lambda-promtail/_index.md

Cloudfront [real-time logs](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/real-time-logs.html) can be sent to a Kinesis data stream. The data stream can be mapped to be an [event source](https://docs.aws.amazon.com/lambda/latest/dg/invocation-eventsourcemapping.html) for lambda-promtail to deliver the logs to Loki.

### Triggering Lambda-Promtail via SQS

> For AWS services supporting sending messages to SQS (ex: S3 with an S3 Notification to SQS), events can be processed through an `SQS queue with lambda trigger` instead of directly configuring the source service to trigger lambda. Lambda-promtail will retrieve the nested events from the SQS messages' body and process them as if them came directly from the source service.

Reviewer: ```suggestion
For AWS services supporting sending messages to SQS (for example, S3 with an S3 Notification to SQS), events can be processed through an `SQS queue with lambda trigger` instead of directly configuring the source service to trigger lambda. Lambda-promtail will retrieve the nested events from the SQS messages' body and process them as if them came directly from the source service.
```

Author: Thanks for your suggestion! I will include it in the PR

### On-Failure log recovery using SQS

> Triggering lambda-promtail through SQS allows handling on-failure recovery of the logs using a secondary SQS queue as a dead-letter-queue. Lambda can be set up such as unsuccessfully processed messages will be sent to the DLQ. After fixing the issue, operators will be able to reprocess the messages by sending back messages from the DLQ to the source queue using the SQS `DLQ redrive` feature.

Reviewer: ```suggestion
Triggering lambda-promtail through SQS allows handling on-failure recovery of the logs using a secondary SQS queue as a dead-letter-queue (DLQ). You can configure lambda so that unsuccessfully processed messages will be sent to the DLQ. After fixing the issue, operators will be able to reprocess the messages by sending back messages from the DLQ to the source queue using the SQS `DLQ redrive` feature.
```

Author: Thanks for your suggestion! I will include it in the PR

Author: Done, also added some links to the AWS documentation
